src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorCompiler.cs

```diff
@@ -1,133 +1,133 @@
-﻿
-// Copyright (c) .NET Foundation. All rights reserved.
+﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
 using System.Linq;
-using Microsoft.AspNetCore.Diagnostics;
+using System.Reflection;
+using System.Runtime.Loader;
+using System.Text;
 using Microsoft.AspNetCore.Mvc.Razor.Compilation;
 using Microsoft.AspNetCore.Razor.Language;
-using Microsoft.AspNetCore.Mvc.Razor.Extensions;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
 
 namespace Microsoft.AspNetCore.Mvc.Razor.Internal
 {
     public class RazorCompiler
     {
-        private readonly ICompilationService _compilationService;
         private readonly ICompilerCacheProvider _compilerCacheProvider;
         private readonly RazorTemplateEngine _templateEngine;
-        private readonly Func<string, CompilerCacheContext> _getCacheContext;
-        private readonly Func<CompilerCacheContext, CompilationResult> _getCompilationResultDelegate;
+        private readonly Func<string, CompilationResult> _getCompilationResult;
+        private readonly ILogger<RazorCompiler> _logger;
+        private readonly CSharpCompiler _compiler;
+        private readonly Action<RoslynCompilationContext> _compilationCallback;
 
         public RazorCompiler(
-            ICompilationService compilationService,
             ICompilerCacheProvider compilerCacheProvider,
-            RazorTemplateEngine templateEngine)
+            RazorTemplateEngine templateEngine,
+            CSharpCompiler compiler,
+            IOptions<RazorViewEngineOptions> optionsAccessor,
+            ILoggerFactory logger)
         {
-            _compilationService = compilationService;
             _compilerCacheProvider = compilerCacheProvider;
             _templateEngine = templateEngine;
-            _getCacheContext = GetCacheContext;
-            _getCompilationResultDelegate = GetCompilationResult;
+            _compiler = compiler;
+            _compilationCallback = optionsAccessor.Value.CompilationCallback;
+            _logger = logger.CreateLogger<RazorCompiler>();
+
+            _getCompilationResult = GetCompilationResult;
         }
 
         private ICompilerCache CompilerCache => _compilerCacheProvider.Cache;
 
         public CompilerCacheResult Compile(string relativePath)
```

Comment:
The goal was to make this one method the user needs to invoke to get caching + compilation. `DefaultRoslynCompilation` required independently caching the result which isn't very obvious.